Changes required for compatability with ESS NeXus files

martyngigg opened the pull request:
**Description of work.**

This replaces #27570. The HDF5 upgrade on Windows was also required for the Python 3 upgrade and was isolated and merged in #27645. This proved that the simple upgrade of HDF5 was not the cause of the build failures on #27570. For clarity a new branch with just the changes required for ESS NeXus files was produced here.

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

**To test:**

<!-- Instructions for testing. -->

*There is no associated issue.*

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

LamarMoore committed:
Changes for hdf5 upgrade compatibility wiht ESS nexus files

LamarMoore committed:
small changes

LamarMoore committed:
add missing algorithm header

LamarMoore committed:
force detectorID sort in NexusGeometryParser

LamarMoore committed:
clang-format

LamarMoore committed:
fix eventid_max calculation in DefaultEventLoader

OwenArnold commented on Framework/DataHandling/src/LoadBankFromDiskTask.cpp:
```diff
@@ -222,10 +235,12 @@ LoadBankFromDiskTask::loadEventId(::NeXus::File &file) {
-      // (from the IDF)
-      // ID. Setting this will abort the loading of the bank.
+    auto offset_min_id = m_min_id + m_loader.pixelID_to_wi_offset;
+    if (offset_min_id >=
```
This addition seems to be responsible for the test failure LoadEventNexusTest (run_MPI_load)

OwenArnold commented on Framework/DataHandling/inc/MantidDataHandling/DefaultEventLoader.h:
```diff
@@ -128,7 +128,7 @@ void DefaultEventLoader::makeMapToEventLists(
     // To avoid going out of range in the vector, this is the MAX INDEX that can
     // go into it
-    eventid_max = static_cast<int32_t>(pixelID_to_wi_vector.size());
+    eventid_max = static_cast<int32_t>(pixelID_to_wi_vector.size() - 1);
```
This change is responsible for the test failure `LoadEventNexusTest::test_load_event_nexus_sans2d_ess`

mantid-builder commented:
@LamarMoore, @martyngigg the build is failing, can you investigate.
<!-- {"problem_type": "failing"} -->

LamarMoore committed:
review changes first pass

mantid-builder commented:
@LamarMoore the build is failing, can you investigate.
<!-- {"problem_type": "failing"} -->

LamarMoore committed:
Fix issue with mpi test in LoadEventNexusTest

LamarMoore committed:
Fix doxygen and unix instability

LamarMoore committed:
doxygen fix

martyngigg commented on Framework/DataHandling/inc/MantidDataHandling/DefaultEventLoader.h:
```diff
@@ -25,59 +25,59 @@ class MANTID_DATAHANDLING_DLL DefaultEventLoader {
 
-  /// True if the event_id is spectrum no not pixel ID
-  bool event_id_is_spec;
+  /// True if the event ID is spectrum no not pixel ID
```
Thank you for cleaning up the variable names.

martyngigg commented on Framework/DataHandling/inc/MantidDataHandling/DefaultEventLoader.h:
```diff
@@ -114,35 +118,34 @@ void DefaultEventLoader::makeMapToEventLists(
-      size_t wi = pixelID_to_wi_vector[j];
+    for (int32_t j = minEventID; j <= m_maxEventID; ++j) {
+      auto index = j + m_detIDtoIndexOffset;
+      size_t wi = m_detIDtoIndexVector[index];
```
Does something guarantee that index is in the the range of `m_detIDtoIndexVector.size()`?

LamarMoore replied:
Yes. The size of the vector is as large as the largest detector ID. The vector is created to improve access speed. See the corresponding method on `MatrixWorkspace` https://github.com/mantidproject/mantid/blob/master/Framework/API/src/MatrixWorkspace.cpp#L540

martyngigg replied:
Ah yes I'd forgotten about the dance this does with those vectors. Thanks.

martyngigg commented on Framework/DataHandling/src/LoadBankFromDiskTask.cpp:
```diff
@@ -55,12 +60,20 @@ void LoadBankFromDiskTask::loadPulseTimes(::NeXus::File &file) {
+  else if (file.hasAttr("start"))
+    file.getAttr("start", thisStartTime);
+  else {
+    auto epoch = DateAndTime(DateAndTime::UNIX_EPOCH);
```
Are there examples where either `offset` or `start` are not defined? Should we `throw` instead as I feel like the file is broken if we don't know the start time...

LamarMoore replied:
👍 will look into this

LamarMoore replied:
Hi Martyn,

This from Matt Jones
```
Hi Lamar, the ESS file writer always records a start time, and I believe the same is true of the ICP at ISIS. Does that answer your question?





9:23
It is probably possible to engineer a file with Mantid that does not contain a start time, but all real, raw data files have one AFAIK.
```

LamarMoore replied:
@martyngigg I could log an error. I am not sure if there is any exception handling for this as it is a task which may be run in several threads.

martyngigg replied:
@LamarMoore Thanks. I think this point is probably too low level to log/throw an exception. We have a check in `LoadEventNexus` (commented on below). As Matt says if we know all files have offsets/start time should we keep throw in the `LoadEventNexus` part if the file doesn't contain it. Or maybe at least a warning that times have been set relative to Unix EPOCH?

martyngigg commented on Framework/DataHandling/src/LoadEventNexus.cpp:
```diff
@@ -420,24 +420,37 @@ void LoadEventNexus::exec() {
+    file.getAttr("start", isooffset);
+    offset = DateAndTime(isooffset);
+  } else
+    offset = DateAndTime(DateAndTime::UNIX_EPOCH);
```
Echoing my comment above. We threw before if we didn't know the offsets. Is it defined for ESS files that if offset is not present then it should be treated as Unix epoch?

LamarMoore replied:
Yes ESS nexus files assume the unix epoch

martyngigg replied:
Ok. Thanks for clarifying.

martyngigg reviewed (commented):
So as I actually opened this I can't mark request changes/approve but I have added some comments. @OwenArnold is probably better placed to review the `NexusGeometry` changes.

OwenArnold commented on Framework/DataHandling/src/BankPulseTimes.cpp:
```diff
@@ -25,7 +26,14 @@ BankPulseTimes::BankPulseTimes(::NeXus::File &file,
+    file.getAttr("offset", startTime);
+  else if (file.hasAttr("start"))
+    file.getAttr("start", startTime);
+  else {
```
This only makes sense if you know that the ESS doens't record offset or start and uses a unix_epoch. Could you please add comment about where this comes from to help future maintainers?

OwenArnold commented on Framework/DataHandling/src/DefaultEventLoader.cpp:
```diff
@@ -110,48 +111,54 @@ DefaultEventLoader::setupChunking(std::vector<std::string> &bankNames,
   return {bank0, bankn};
 }
 
+std::pair<int32_t, int32_t> DefaultEventLoader::getMinMaxDetID() {
```
Could be `const` method I think.

OwenArnold commented on Framework/DataHandling/src/LoadBankFromDiskTask.cpp:
```diff
@@ -55,12 +60,20 @@ void LoadBankFromDiskTask::loadPulseTimes(::NeXus::File &file) {
+    file.getAttr("offset", thisStartTime);
+  else if (file.hasAttr("start"))
+    file.getAttr("start", thisStartTime);
+  else {
```
Same as code in `BankPulseTimes.cpp`, minor point - consider refactoring. but also needs same comment.

OwenArnold commented on Framework/DataHandling/src/LoadBankFromDiskTask.cpp:
```diff
@@ -350,12 +374,13 @@ void LoadBankFromDiskTask::run() {
     // Open the bankN_event group
-    file.openGroup(entry_name, entry_type);
+    if (NexusDescriptor::findAndOpenParentGroup(file, entry_type, entry_name))
+      file.openGroup(entry_name, entry_type);
```
Is this still safe? If we can't open the parent group, we can still safely continue?

martyngigg replied:
This could be quite slow for files with many entries as `getEntries()` is slow. Can the caller not verify the file looks like an expected file structure and pass this in. For SNS/ISIS we just hardcode the path in the file. Is the ESS structure more fluid?

OwenArnold commented on Framework/DataHandling/src/LoadBankFromDiskTask.cpp:
```diff
@@ -412,12 +437,12 @@ void LoadBankFromDiskTask::run() {
-    m_loader.alg->getLogger().error() << e.what() << '\n';
+    m_loader.m_loadAlgorithm->getLogger().error() << e.what() << '\n';
     m_loadError = true;
   } catch (...) {
```
This is a bit of a segway, but is there a possibility of the nexus api propagating H5 exceptions? In which case falls into (...). Note that in H5cpp, H5Exception does not have std::exception as base https://support.hdfgroup.org/HDF5/doc/cpplus_RM/_h5_exception_8h_source.html

LamarMoore replied:
The SNS have plans to rewrite this using H5 consistently. May not be worth the effort.

OwenArnold commented on Framework/DataHandling/src/LoadEventNexus.cpp:
```diff
@@ -420,24 +420,37 @@ void LoadEventNexus::exec() {
+    std::string message =
+        "event_time_zero has zero dimension. Cannot establish run "
+        "start.";
+    logger.warning(message);
```
Why a warning rather than an error given that you throw on next line?

OwenArnold commented on Framework/DataHandling/src/LoadEventNexus.cpp:
```diff
@@ -420,24 +420,37 @@ void LoadEventNexus::exec() {
+  } else if (file.hasAttr("start")) {
+    file.getAttr("start", isooffset);
+    offset = DateAndTime(isooffset);
+  } else
```
Another repeated block similar to the two above

OwenArnold commented on Framework/DataHandling/src/LoadEventNexus.cpp:
```diff
@@ -447,33 +460,23 @@ firstLastPulseTimes(::NeXus::File &file, Kernel::Logger &logger) {
+  } else {
+    std::string message = "Unrecognised type for event_time_zero";
+    logger.warning(message);
+    throw std::runtime_error(message);
```
Warning then throw.
Also is it possible to report what the actual held type is. That would help future maintainers.

OwenArnold commented on Framework/DataHandling/src/LoadEventNexus.cpp:
```diff
@@ -821,8 +824,16 @@ void LoadEventNexus::loadEvents(API::Progress *const prog,
   auto firstPulseT = DateAndTime::maximum();
+
+  if (!NexusDescriptor::findAndOpenParentGroup(*m_file, classType))
+    g_log.warning()
```
This warning might solve my question above. In which case you could just add a comment that no event data would be loaded.

OwenArnold commented on Framework/DataHandling/src/ProcessBankData.cpp:
```diff
@@ -159,14 +159,17 @@ void ProcessBankData::run() { // override {
       // the event
       const detid_t detId = event_id[eventIndex];
       if (detId >= m_min_id && detId <= m_max_id) {
+        auto index = m_loader.m_isEventIDSpec
```
👍

OwenArnold commented on Framework/Kernel/src/NexusDescriptor.cpp:
```diff
@@ -271,5 +274,37 @@ void NexusDescriptor::walkFile(::NeXus::File &file, const std::string &rootPath,
+      return true;
+    } else {
+      for (const auto &[name, type] : entries) {
+        try {
```
I'm guessing the try-catch is there because somethings are not groups. Is there a better way to identify aprori `is_group` etc? If not, can we be more specific about the exception type - this could cause future maintenance problems if it's casting too much of a wide net.

OwenArnold commented on Framework/NexusGeometry/inc/MantidNexusGeometry/TubeHelpers.h:
```diff
@@ -27,9 +27,9 @@ MANTID_NEXUSGEOMETRY_DLL std::vector<detail::TubeBuilder>
+MANTID_NEXUSGEOMETRY_DLL std::vector<size_t>
 notInTubes(const std::vector<detail::TubeBuilder> &tubes,
-           std::vector<Mantid::detid_t> detIDs);
+           const std::vector<Mantid::detid_t> &detIDs);
```
👍

OwenArnold commented on Framework/NexusGeometry/src/TubeHelpers.cpp:
```diff
@@ -8,6 +8,7 @@
 #include "MantidGeometry/Objects/IObject.h"
 #include <algorithm>
 #include <iterator>
+#include <tbb/parallel_sort.h>
```
Is this include still needed?

OwenArnold reviewed (changes requested):
Looks good. I've suggested a few things for clean-up.

LamarMoore committed:
review changes

OwenArnold commented on Framework/Nexus/src/NexusIOHelper.cpp:
```diff
@@ -0,0 +1,33 @@
+                       "Assuming Unix epoch. \n";
+  }
+  return startTime;
+}
```
👍

LamarMoore committed:
fix linking error

LamarMoore committed:
fix file name case

mantid-builder commented:
@LamarMoore the build is failing, can you investigate.
<!-- {"problem_type": "failing"} -->

mantid-builder commented:
@LamarMoore have you had a chance to see why the build is failing?
<!-- {"problem_type": "failing"} -->

martyngigg committed:
Move header implementation to cpp file.

Anonymous namespaces imply they should be restricted
to a single TU but being in a header file is in contrast
with this. Implement in a .cpp file for regular sharing.

martyngigg committed:
Clean up code with more modern constructs

martyngigg committed:
Swap out Poco::Process for boost::process

Shared memory is provided by boost. To avoid mixing too
many libraries in the implementation use boost to start
child processes.

martyngigg commented:
Converted to draft to remove from ready to review list.
